tests/test_apel_plugin.py

```diff
@@ -21,3 +25,90 @@ async def test_regex_dict_lookup(self):
             await functions.regex_dict_lookup(term_c, dict)
         assert pytest_error.type == SystemExit
         assert pytest_error.value.code == 1
+
+    async def test_get_begin_previous_month(self):
+        time_a = datetime(2022, 10, 23, 12, 23, 55)
+        time_b = datetime(1970, 1, 1, 00, 00, 00)
+
+        result = await functions.get_begin_previous_month(time_a)
+        assert result == datetime(2022, 9, 1, 00, 00, 00, tzinfo=pytz.utc)
+
+        result = await functions.get_begin_previous_month(time_b)
+        assert result == datetime(1969, 12, 1, 00, 00, 00, tzinfo=pytz.utc)
+
+    async def test_sql_filter(self):
+        create_table_sql = """
+                           CREATE TABLE IF NOT EXISTS records(
+                           site TEXT NOT NULL,
+                           year INTEGER NOT NULL,
+                           month INTEGER NOT NULL
+                           )
+                           """
+
+        insert_record_sql = """
+                            INSERT INTO records(
+                               site,
+                               year,
+                               month
+                           )
+                           VALUES(
+                               ?, ?, ?
+                           )
+                           """
+
+        conn = await aiosqlite.connect(":memory:")
+        cur = await conn.cursor()
+        await cur.execute(create_table_sql)
+
+        site_list = ["site_a", "site_b", "site_c"]
+        year_list = [2020, 2021, 2022]
+        month_list = [1, 7, 12]
+
+        for site in site_list:
+            for year in year_list:
+                for month in month_list:
+                    data_tuple = (site, year, month)
+                    await cur.execute(insert_record_sql, data_tuple)
+        await conn.commit()
+        await cur.close()
+
+        for site in site_list:
+            for year in year_list:
+                for month in month_list:
+                    dst = await aiosqlite.connect(":memory:")
```

Comment:
I want to repeat the test for several values of `site`, `month`, and `year`. If I wouldn't backup the DB, I would have to recreate the initial DB for every test, since the filter removes entries.
But maybe you know an alternative to this?

Reply:
I guess for a test this is fine. I assumed that you would use the same pattern in your code as well. I'm still not a huge fan of all this. I haven't looked into the specifics, but it seems to me that this (or most of this)

```python
summary_db = await functions.create_summary_db(config, records)
filtered_db = await functions.sql_filter(summary_db, month, year, site)
grouped_summary_list = await functions.group_summary_db(filtered_db)
summary = await functions.create_summary(grouped_summary_list)
```
can be done with two SQL queries: One for insertion, one for querying the data in your preferred format. It is rather uncommon to juggle around actual (filtered) databases rather than immutable views on a single database.

Reply:
I just wanted to work on this and move the filtering into `group_summary_db` (without deleting from the db) as discussed. But I just noticed that I only need the filtering for the `republish` script, not for the usual publishing. This means that

1. I would need some case differentiation when calling `group_summary_db` (publish vs. republish)
2. `group_summary_db` would need additional arguments: `month`, `year`, and  `site` (that are not used during usual publishing)

So I'm not sure if this is still a good idea. What do you think?

Reply:
Maybe an optional parameter `filter_by` (or something like that) of type `(int, int, str)` (which represents (year, month, site)) would be an option? I would default this optional parameter to `None`. So something along the lines of:

```python
def group_summary_db(filter_by: (int, int, str) = None):
    # ... 
```

I'm not sure if this is syntactically correct though.

Reply:
Done with commit [a12153c](https://github.com/ALU-Schumacher/AUDITOR-APEL-plugin/pull/22/commits/a12153c79b8b95a651fbfc626de4956005f498d1)
Does this look ok?

Reply:
:+1: